29 issues 24 Add EF and Business logic HabitsBookMicroService

AATerekhov opened the pull request:
Steps before launching containers

AATerekhov committed:
Add business logic of the card

AATerekhov committed:
Added the implementation of the habits application service

AATerekhov committed:
added the implementation of the application coins service

AATerekhov committed:
Add EF Migration

YauhenIlyushenka commented on BookOfHabits/BookOfHabits/Controllers/CardsController.cs:
```diff
@@ -0,0 +1,55 @@
+
+        [HttpPut]
+        public async Task UpdateCardAsync(UpdateCardRequest request)
+        {
```
if we have one instruction only it would be great to use => instead of "{return ...}"

AATerekhov replied:
I think it's too early to wipe your tails =)

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Application.Service/CardsApplicationService.cs:
```diff
@@ -0,0 +1,104 @@
+        {
+            var card = await cardRepository.GetByIdAsync(x => x.Id.Equals(cardInfo.Id), cancellationToken: token)
+                 ?? throw new NotFoundException(FormatFullNotFoundErrorMessage(cardInfo.Id, nameof(Card)));
+            if (cardInfo.Name is not null)
```
I have no clue about logic, but the multitude of 'if' it seems like 'smell code' and should be refactored

AATerekhov replied:
There is no logic here, it is just allowed that there is no data in request entity.
I couldn't find a quick, more concise solution to protect against data erasure if null comes from the front.
I accept recommendations.

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Application.Service/CardsApplicationService.cs:
```diff
@@ -0,0 +1,104 @@
+                ?? throw new NotFoundException(FormatFullNotFoundErrorMessage(cardId, nameof(Card)));
+            var template = await templateValuesRepository.GetByIdAsync(x => x.Id.Equals(card.TemplateValues.Id), cancellationToken: token)
+                ?? throw new NotFoundException(FormatFullNotFoundErrorMessage(card.TemplateValues.Id, nameof(TemplateValues)));
+            if (tempateValuesInfo.Status is not null)
```
I have no clue about logic, but the multitude of 'if' it seems like 'smell code' and should be refactored

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Application.Service/HabitsApplicationService.cs:
```diff
@@ -0,0 +1,146 @@
+                                  isEndless: true,
+                                  durationTime: 3600);
+
+            var repetition = new Repetition(maxCountPositive: 5,
```
we should get rid of 'magic numbers' from the our code and instead of it we can use constants

AATerekhov replied:
Default objects are created here, ideally to remove them in extensions, but for now it is not important. It will be removed in the future.

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Application.Service/HabitsApplicationService.cs:
```diff
@@ -0,0 +1,146 @@
+                                            countLimit: 10);
+
+            var timeResetInterval = new TimeResetInterval(options: ResetIntervalOptions.EveryDay,
+                                                          timeTheDay: 43_200,
```
the same

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/CoinsOptions.cs:
```diff
@@ -13,6 +13,7 @@ public enum CoinsOptions
         CheckingLimit = 4,
-        CheckingReport = 8
+        CheckingReport = 8,
+        All = 15
```
why did you setup 15 after 8? It seems like it should be 16 if you use bit flags instead of a standard enum

VladimirFKulagin replied:
you can use any combination of bits and it looks correct for "All"

AATerekhov replied:
All - this is an auxiliary field, which value to expect if all flags are enabled.
1 + 2 +4 + 8 = 15

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/HabitOptions.cs:
```diff
@@ -13,5 +13,6 @@ public enum HabitOptions
         Delayed =1,
         Reset = 2,
         Repetition = 4,
+        All = 7,
```
the same

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/ResetIntervalOptions.cs:
```diff
@@ -0,0 +1,18 @@
+        EveryDay = 1,
+        Weekday = 2,
+        OnceAMonth = 4,
+        All = 7
```
the same

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/WeekDays.cs:
```diff
@@ -16,6 +16,7 @@ public enum WeekDays
         Saturday = 32,
-        Sunday = 64        
+        Sunday = 64,
+        All = 127
```
the same

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Application.Service/InstallCardApplicationService.cs:
```diff
@@ -0,0 +1,34 @@
+            Person owner = await personRepository.GetByIdAsync(x => x.Id.Equals(installCardModel.PersonId), cancellationToken: token)
+                ?? throw new NotFoundException(FormatFullNotFoundErrorMessage(installCardModel.PersonId, nameof(Person)));
+
+            Habit habit = await habitRepository.GetByIdAsync(x => x.Id.Equals(installCardModel.HabitId), includes: $"{nameof(Habit.Owner)};{nameof(Habit.Card)}", cancellationToken: token)
```
did you check this one and it works properly? In implementation you have mentioned separator like ',', but here you have made separator between entities like ';'

AATerekhov replied:
Thanks for the tip, there's a mistake.

AATerekhov committed:
Edited includes Repository method in Services.

AATerekhov committed:
Added default Values Models and consts in the Domain models.

AATerekhov committed:
A Card validator has been added. Removed multiple IF

Ainur555 commented on BookOfHabits/BookOfHabits/Controllers/CoinsController.cs:
```diff
@@ -0,0 +1,49 @@
+                                 IChooseHabitApplicationService chooseHabitApplicationService,
+                                 IMapper mapper) : ControllerBase
+    {
+        [HttpGet("room/{roomId:guid}")]
```
add routes for your methods and comments. For example [HttpPut("CreateCoins")]

AATerekhov replied:
I prefer minimal routes. if there are enough controller names and it's unique, I'll leave it that way.

Ainur555 commented on BookOfHabits/BookOfHabits/Requests/Person/CreatePersonRequest.cs:
```diff
@@ -0,0 +1,8 @@
+    public class CreatePersonRequest
+    {
+        public Guid Id { get; init; }
+        public required string Name { get; init; }
```
add [Require] attribute, where it need

AATerekhov replied:
This is a validation attribute. I don't like using attributes, I use Fluentvalidation

AATerekhov commented:
Do we even have an agreement, how long will this hang here?

YauhenIlyushenka commented on BookOfHabits/BookOfHabits/Infrastructure/Validators/CreateCardValidator.cs:
```diff
@@ -0,0 +1,21 @@
+                .NotNull();
+        }
+        internal static bool ValidateOptionsField(CardOptions options)
+            => (int)options <= (int)CardOptions.All ? true : false;
```
it's a smell code. Should be refactored without part of "? true: false"

YauhenIlyushenka commented on BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Propertys/Delay.cs:
```diff
@@ -4,29 +4,33 @@ namespace BookOfHabitsMicroservice.Domain.Entity.Propertys
-        public DateTime Finish { get; }
-        public bool IsEndless { get; }
-        public Delay(Guid id, Habit habit, bool isAfterATime, long afterTime, DateTime start, DateTime finish, bool isEndless) 
+        readonly int _ticksSec = 10_000_000;
```
the fields int_ticksSec and int_day are had in TimeResetInterval model as well with the same values.

I have no clue about the logic, but it seem like code duplication. Because if I decide to change it, I will be forced to find all places with this variables. Instead of it I can determine constants and use it in the both models so it will be great solution to simplify  understanding code and make some cahnges in it in the long run.

AATerekhov committed:
Fixed business logic errors
Added descriptions of the enumerations
Changing entities is done by the full UpdateEntityModel
